### PR TITLE
check for segmentation fault when executing instructions

### PR DESCRIPTION
**Pull request by Dylan-Brotherston**

### CODE REVIEW

**Comment by Dylan-Brotherston** [comment]
This commit causes mipsy-lib to emit a runtime error when instructions are executed outside of TEXT or KTEXT.

The error message text itself could be improved to provide more useful information.

**Comment by Dylan-Brotherston** [comment]
@SpanishPear 35c0c32 also touches mips-web as it removes the need for the `UnknownSyscall` pseudo-syscall. You might want to double check this commit.

File: crates/mipsy_lib/src/error/runtime/mod.rs [hunk 1]
```diff
@@ -684,6 +698,211 @@ impl Error {
...
+                    ));
+                    error.push('\n');
+                }
+
+                let decompiled_all = decompile::decompile_into_parts(binary, inst_set);
+                let decompiled_next = decompile::decompile_inst_into_parts(binary, inst_set, inst, state.pc() + 4);
+                if decompiled.location.is_none() || ( decompiled_all.contains_key(&decompiled_next.addr) && decompiled_next.location.is_none()) {
+                    // The current instruction should always be `syscall`, so we can't get a pseudo-instruction
+                    // Therefor, we don't need to expand it.
+                    unreachable!()
```
Comment (insou22) [inline crates/mipsy_lib/src/error/runtime/mod.rs:746]: this argument is faulty -- `syscall` could be *part of* a pseudo-instruction (like in the dbg_print instns, although they cannot fail in this way) -- so this is making an assumption on the set of pseudoinstns

**Review by insou22 — COMMENTED** [review]
I do think the unimplementeds ought to be up to the mipsy frontend to deal with -- I'd rather deal with the unimplemented file stuff by just implementing it properly in mipsy (won't be very hard), but I imagine they're going to stay unimplemented in mipsy_web

File: crates/mipsy_lib/src/error/runtime/mod.rs (hunk 1 above)
Comment (Dylan-Brotherston) [reply crates/mipsy_lib/src/error/runtime/mod.rs:746]: That is a very good point.
I will add the normal expansion code.

**Comment by SpanishPear** [comment]
Does this mean that mipsy_web will now/still panic upon invalid syscalls?

**Comment by Dylan-Brotherston** [comment]
This will raise a runtime error from mipsy_lib just like any other error. So I think mipsy Lin should handle this correctly as is.

**Comment by Dylan-Brotherston** [comment]
The syscall error handling and the better segment semantics should stop two of the common causes of panics, so should make things better for mipsy_web.

**Review by SpanishPear — APPROVED** [review]

**Comment by Dylan-Brotherston** [comment]
@insou22 The segfault error could be better but it's good enough to help students as it currently is.

So this is really to merge if you are happy.

File: crates/mipsy_lib/src/error/runtime/mod.rs [hunk 2]
```diff
@@ -840,39 +1217,39 @@ impl Error {
...
-                    }
-                    _ if addr >= heap_stack_middle && addr < STACK_BOT => {
-                        vec![format!("the minimum stack address is `{}{}`", "0x".bold(), format!("{:08x}", STACK_BOT).bold())]
-                    }
-                    _ if addr > STACK_TOP => {
-                        vec![format!("the maxmimum stack address is `{}{}`", "0x".bold(), format!("{:08x}", STACK_TOP).bold())]
-                    }
-                    // hopefully unreachable
-                    _ => vec![]
+                match get_segment(addr) {
```
Comment (insou22) [inline crates/mipsy_lib/src/error/runtime/mod.rs:1228]:
```
match (addr) {
    // special cases
    0 => // null
    ...? => //
    _ => match get_segment(addr) {
        // code goes here
    }
}
```

File: crates/mipsy_lib/src/runtime/mod.rs [hunk 3]
```diff
@@ -269,38 +284,38 @@ impl Runtime {
...
                             });
                             self.timeline.state_mut().write_register(Register::V0.to_u32(), n_bytes);
-                            
+                
                             self
                         })
                     )
                 }
-                SYS15_WRITE => {
+                SYS15_WRITE =>{
```
Comment (insou22) [inline crates/mipsy_lib/src/runtime/mod.rs:306]: angery

**Review by insou22 — CHANGES_REQUESTED** [review]

**Review by insou22 — APPROVED** [review]

**Pull request merged**